system/Database/BaseBuilder.php

```diff
@@ -1854,6 +1855,108 @@ public function setData($set, ?bool $escape = null, string $alias = '')
         return $this;
     }
 
+    /**
+     * Compiles an upsert query and returns the sql
+     *
+     * @return string
+     *
+     * @throws DatabaseException
+     */
+    public function getCompiledUpsert()
+    {
+        [$currentTestMode, $this->testMode] = [$this->testMode, true];
+
+        $sql = implode(";\n", $this->upsert());
+
+        $this->testMode = $currentTestMode;
+
+        return $this->compileFinalQuery($sql);
+    }
+
+    /**
+     * Converts call to batchUpsert
+     *
+     * @param array|object|null $set
+     *
+     * @return false|int|string[] Number of affected rows or FALSE on failure, SQL array when testMode
+     *
+     * @throws DatabaseException
+     */
+    public function upsert($set = null, ?bool $escape = null)
+    {
+        if ($set === null) {
+            $set = empty($this->binds) ? null : [array_map(static fn ($columnName) => $columnName[0], $this->binds)];
```

Comment:
This retrieve only values from the `$this->binds`.
And calling `$this->setData($set, $escape)` below sets the escape values with the parameter `$escape`.

The first escape values set are overridden with the parameter value.